pkg/kubelet/apis/pluginregistration/v1alpha/api.proto

```diff
@@ -0,0 +1,37 @@
+// To regenerate api.pb.go run hack/update-generated-kubelet-plugin-registration.sh
+syntax = 'proto3';
+
+package pluginregistration;
+
+import "github.com/gogo/protobuf/gogoproto/gogo.proto";
+
+option (gogoproto.goproto_stringer_all) = false;
+option (gogoproto.stringer_all) =  true;
+option (gogoproto.goproto_getters_all) = true;
+option (gogoproto.marshaler_all) = true;
+option (gogoproto.sizer_all) = true;
+option (gogoproto.unmarshaler_all) = true;
+option (gogoproto.goproto_unrecognized_all) = false;
+
+message InfoResponse {
+	// Type of the Plugin. CSIPlugin or DevicePlugin
+	string type = 1;
+	// Plugin name.
+	string name = 2;
+        // Plugin service API version
+	string version = 3;
+}
+
+message Empty {
+}
+
+message RegistrationStatus {
+	bool plugin_registered  = 1;
+	string error  = 2;
+}
+
+// Identity is the service advertised by the Plugins
+service Identity {
+	rpc Info(Empty) returns (InfoResponse) {}
+	rpc UpdateRegistrationStatus(RegistrationStatus) returns (Empty) {}
+}
```

Comment:
I saw you defined two rpc calls to finish the three-way handshaking. I think an alternative way to implement three-way handshaking is to use bi-directional grpc calls with something like below:

service Registration {
  rpc Register(stream RegistrationStatus) returns (stream PluginInfo) {}
}

// On plugin side
func (s *registrationServer) Register(stream pb.Registration_RegisterServer) error {
    if in, err := stream.Recv(); err != nil {
      return err
    }
    rqst := makeRegistrationRequest()  // fill in plugin info
    if err := stream.Send(rqst); err != nil {
      return err
    }
    if in, err := stream.Recv(); err != nil {
      return err
    }
    if !in.PluginRegistered {
      return fmt.Errorf("Registration failed: %s", in.Error)
    }
    return nil
}

// On kubelet side
client := pb.NewRegistrationClient(conn)
if stream, err := client.Register(context.Background()); err != nil {
  return err
}
defer stream.CloseSend()
resp := new(pb.RegistrationStatus)
// Sends an empty stream request to initialize handshaking
if err := stream.Send(resp); err != nil {
  return err
}
rqst, err := stream.Recv()
// calls handler callback to verify registration request
if verifyErr = verifyRegistration(rqst); verifyErr != nil {
  resp.PluginRegistered = false
  resp.error = verifyErr.Error()
} else {
  resp.PluginRegistered = true
}
if err := stream.Send(resp); err != nil {
  return err
}
return verifyErr

I am actually not sure which approach is better. The way we use bi-directional grpc is also not very intuitive as the handshaking is initiated with an empty RegistrationStatus. The advantage I can see is that the handshaking is performed in a single grpc call and it perhaps easier for plugins to implement things in the right way.